### PR TITLE
[TACHYON-699] Add FileUtils.delete() and test cases

### PR DESCRIPTION
**Pull request by miasantreble**
https://tachyon.atlassian.net/browse/TACHYON-699

### CODE REVIEW

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java [hunk 1]
```diff
@@ -70,4 +70,22 @@ public void changeLocalDirPermissionTests() throws IOException {
...
+    Assert.assertTrue(toFile.exists());
+  }
+
+  @Test
+  public void deleteFileTest() throws IOException {
+    File toDelete = mTestFolder.newFile("toDelete.txt");
+    // Delete a file and verify
+    FileUtils.delete(toDelete);
+    Assert.assertFalse(toDelete.exists());
+  }
```
Comment (apc999) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: Could you add another test case that tries to delete a non-existing file and expect an `IOException`?

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java [hunk 2]
```diff
@@ -70,4 +70,22 @@ public void changeLocalDirPermissionTests() throws IOException {
...
+    File toFile = mTestFolder.newFile("to.txt");
+    // Move a file and verify
+    FileUtils.move(fromFile, toFile);
+    Assert.assertFalse(fromFile.exists());
+    Assert.assertTrue(toFile.exists());
+  }
+
+  @Test
+  public void deleteFileTest() throws IOException {
+    File toDelete = mTestFolder.newFile("toDelete.txt");
```
Comment (apc999) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: In this testcase, could you create a temp dir and delete it ?

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java (hunk 1 above)
Comment (miasantreble) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: @apc999 looks like deleting a non-existent file doesn't result in IOException, http://docs.oracle.com/javase/7/docs/api/java/io/File.html#delete%28%29

Comment (miasantreble) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: new test case added

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java (hunk 2 above)
Comment (miasantreble) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: done

File: common/src/main/java/tachyon/util/io/FileUtils.java [hunk 3]
```diff
@@ -152,4 +152,17 @@ public static void createBlockPath(String path) throws IOException {
   public static void move(File from, File to) throws IOException {
     Files.move(from, to);
   }
+
+  /**
+   * Delete the file or directory
```
Comment (uronce-cc) [inline common/src/main/java/tachyon/util/io/FileUtils.java]: If the directory is not empty, should this method succeed? If it can not succeed, `File.delete` can satisfy the requirements, but I think for this utility method, we should succeed?

Comment (apc999) [inline common/src/main/java/tachyon/util/io/FileUtils.java]: I think either is fine as long as we specify those corner cases clearly in javadoc.

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java [hunk 4]
```diff
@@ -70,4 +70,32 @@ public void changeLocalDirPermissionTests() throws IOException {
...
+    File toFile = mTestFolder.newFile("to.txt");
+    // Move a file and verify
+    FileUtils.move(fromFile, toFile);
+    Assert.assertFalse(fromFile.exists());
+    Assert.assertTrue(toFile.exists());
+  }
+
+  @Test
+  public void deleteFileTest() throws IOException {
+    File tempFile = mTestFolder.newFile("toDelete.txt");
```
Comment (apc999) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: NIT: to keep the file name and folder name both consistent, maybe 
```java
File tempFile = mTestFolder.newFile("fileToDelete");
File tempFolder = mTestFolder.newFolder("tdirToDelete");
```

File: common/src/main/java/tachyon/util/io/FileUtils.java [hunk 5]
```diff
@@ -152,4 +152,17 @@ public static void createBlockPath(String path) throws IOException {
...
+   *
+   * Current implementation uses {@link java.io.File#delete();}, may change if
+   * there is a better solution.
+   *
+   * @param file file to delete
+   * @throws IOException when fails to delete
+   */
+  public static boolean delete(File file) throws IOException {
+    return file.delete();
```
Comment (apc999) [inline common/src/main/java/tachyon/util/io/FileUtils.java]: The javadoc is inconsistent with how the method behaves. When it failed to delete this file, this method returns `false` rather than throwing an exception. Also `file.delete()` actually not throwing `IOException` but `SecurityException`. 

Here is my suggestion: I think it is fine to keep the signature to throw `IOException`, but changing the returned type from `boolean` to `void` and if `file.delete()` fails, throw `IOException`? That is more aligned to java nio and in the long run, we will re-implement this method using java nio package, the reason we are still not doing so is because nio is not included in java6 and Tachyon is still supporting java6.

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java [hunk 6]
```diff
@@ -70,4 +70,32 @@ public void changeLocalDirPermissionTests() throws IOException {
...
+    File tempFolder = mTestFolder.newFolder("temp-dir");
+    // Delete a file and a directory
+    Assert.assertTrue(FileUtils.delete(tempFile));
+    Assert.assertTrue(FileUtils.delete(tempFolder));
+    Assert.assertFalse(tempFile.exists());
+    Assert.assertFalse(tempFolder.exists());
+  }
+
+  @Test
+  public void deleteNoExistentFileTest() throws IOException {
```
Comment (apc999) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: could you also add a corresponding test for dir not existing?

Comment (miasantreble) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: I think ghostFile is an empty File instance which cannot tell the difference of file or dir until createNewFile() or mkdir() is called?

**Comment by gpang** [comment]
Does anything use FileUtils.delete()?

**Comment by apc999** [comment]
@gpang , currently no, In the future, we want to use `FileUtils` to abstract operations in  `TieredBlockStore` when moving or removing a block, any maybe also other places like `LocalFileBlockWriter`.

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java (hunk 6 above)
Comment (apc999) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: OK that makes sense. Thanks.

File: common/src/main/java/tachyon/util/io/FileUtils.java (hunk 5 above)
Comment (miasantreble) [inline common/src/main/java/tachyon/util/io/FileUtils.java]: done, thanks for the suggestion

File: common/src/test/java/tachyon/util/io/FileUtilsTest.java (hunk 6 above)
Comment (gpang) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: Could you change "NoExistent" -> "NonExistent"? Thanks.

Comment (miasantreble) [inline common/src/test/java/tachyon/util/io/FileUtilsTest.java]: Done

**Comment by apc999** [comment]
+1 LGTM

**Comment by gpang** [comment]
LGTM
